logstash-core/lib/logstash/agent.rb

```diff
@@ -31,6 +38,18 @@ class LogStash::Agent
   #   :auto_reload [Boolean] - enable reloading of pipelines
   #   :reload_interval [Integer] - reload pipelines every X seconds
   def initialize(settings = LogStash::SETTINGS, source_loader = nil)
+
+    # Check that we are the sole live instance
+    if (CURRENT.compare_and_set(nil, self))
+      begin
+        raise "fake"
+      rescue => e
+        CURRENT_INITIALIZED_BACKTRACE.set(e.backtrace);
```

Comment:
We can get the backtrace and avoid the overhead of raise/rescue with [Kernel#caller](https://ruby-doc.org/core-2.5.1/Kernel.html#method-i-caller), which also allows us to skip the current frame which will always be the same:

~~~ ruby
if (CURRENT.compare_and_set(nil,self))
  backtrace = caller(1) # don't include the current frame
  CURRENT_INITIALIZED_BACKTRACE.set(backtrace)
else
  raise # ...
end
~~~